// @vicinae/api/dist/api/proto/api.js
"use strict";
// generated by vicinae figura codegen - do not edit manually
Object.defineProperty(exports, "__esModule", { value: true });
exports.Client = exports.RpcTransport = void 0;
class RpcTransport {
    transport;
    constructor(transport) {
        this.transport = transport;
    }
    dispatchMessage(data) {
        const msg = JSON.parse(data);
        if (msg.id !== undefined) {
            const handler = this.requestMap.get(msg.id);
            if (handler) {
                if (msg.error)
                    handler.reject(msg.error);
                else
                    handler.resolve(msg.result);
                this.requestMap.delete(msg.id);
            }
        }
        if (msg.id === undefined && msg.method) {
            for (const cb of this.handlers.get(msg.method) ?? []) {
                cb(msg.params);
            }
        }
    }
    request(method, params) {
        const id = this.id++;
        const promise = new Promise((resolve, reject) => {
            this.requestMap.set(id, { resolve: (msg) => resolve(msg), reject });
        });
        this.sendMessage({ jsonrpc: "2.0", id, method, params });
        return promise;
    }
    subscribe(method, cb) {
        const handlers = this.handlers.get(method);
        if (handlers)
            handlers.push(cb);
        else
            this.handlers.set(method, [cb]);
        return {
            unsubscribe: () => {
                const handlers = this.handlers.get(method);
                if (handlers) {
                    handlers.splice(handlers.indexOf(cb), 1);
                }
            },
        };
    }
    sendMessage(msg) {
        this.transport.send(JSON.stringify(msg));
    }
    id = 1;
    requestMap = new Map();
    handlers = new Map();
}
exports.RpcTransport = RpcTransport;
class ApplicationService {
    transport;
    constructor(transport) {
        this.transport = transport;
    }
    list(target) {
        return this.transport.request("Application/list", { target });
    }
    open(target, appId) {
        return this.transport.request("Application/open", { target, appId });
    }
    getDefault(target) {
        return this.transport.request("Application/getDefault", { target });
    }
    showInFileBrowser(target, select) {
        return this.transport.request("Application/showInFileBrowser", {
            target,
            select,
        });
    }
    runInTerminal(opts) {
        return this.transport.request("Application/runInTerminal", { opts });
    }
}
class UIService {
    transport;
    constructor(transport) {
        this.transport = transport;
    }
    render(json) {
        return this.transport.request("UI/render", { json });
    }
    showToast(id, title, message, style) {
        return this.transport.request("UI/showToast", {
            id,
            title,
            message,
            style,
        });
    }
    updateToast(id, title) {
        return this.transport.request("UI/updateToast", { id, title });
    }
    hideToast(id) {
        return this.transport.request("UI/hideToast", { id });
    }
    showHud(text, clear_root, popToRoot) {
        return this.transport.request("UI/showHud", {
            text,
            clear_root,
            popToRoot,
        });
    }
    closeMainWindow(clearRoot, popToRoot) {
        return this.transport.request("UI/closeMainWindow", {
            clearRoot,
            popToRoot,
        });
    }
    popToRoot(clearSearchBar) {
        return this.transport.request("UI/popToRoot", { clearSearchBar });
    }
    confirmAlert(payload) {
        return this.transport.request("UI/confirmAlert", { payload });
    }
    pushView() {
        return this.transport.request("UI/pushView", {});
    }
    popView() {
        return this.transport.request("UI/popView", {});
    }
    setSearchText(text) {
        return this.transport.request("UI/setSearchText", { text });
    }
    getSelectedText() {
        return this.transport.request("UI/getSelectedText", {});
    }
    sendDesktopNotification(data) {
        return this.transport.request("UI/sendDesktopNotification", { data });
    }
    viewPoped(handler) {
        return this.transport.subscribe("UI/viewPoped", (msg) => handler());
    }
    viewPushed(handler) {
        return this.transport.subscribe("UI/viewPushed", (msg) => handler());
    }
}
class WindowManagementService {
    transport;
    constructor(transport) {
        this.transport = transport;
    }
    focusWindow(winId) {
        return this.transport.request("WindowManagement/focusWindow", { winId });
    }
    getActiveWindow() {
        return this.transport.request("WindowManagement/getActiveWindow", {});
    }
    getActiveWorkspace() {
        return this.transport.request("WindowManagement/getActiveWorkspace", {});
    }
    getWindows(workspaceId) {
        return this.transport.request("WindowManagement/getWindows", {
            workspaceId,
        });
    }
    getScreens() {
        return this.transport.request("WindowManagement/getScreens", {});
    }
    getWorkspaces() {
        return this.transport.request("WindowManagement/getWorkspaces", {});
    }
    setWindowBounds(winId, bounds) {
        return this.transport.request("WindowManagement/setWindowBounds", {
            winId,
            bounds,
        });
    }
}
class ClipboardService {
    transport;
    constructor(transport) {
        this.transport = transport;
    }
    copy(content, options) {
        return this.transport.request("Clipboard/copy", { content, options });
    }
    paste(content) {
        return this.transport.request("Clipboard/paste", { content });
    }
    clear() {
        return this.transport.request("Clipboard/clear", {});
    }
    readContent() {
        return this.transport.request("Clipboard/readContent", {});
    }
}
class StorageService {
    transport;
    constructor(transport) {
        this.transport = transport;
    }
    get(key) {
        return this.transport.request("Storage/get", { key });
    }
    set(key, value) {
        return this.transport.request("Storage/set", { key, value });
    }
    remove(key) {
        return this.transport.request("Storage/remove", { key });
    }
    clear() {
        return this.transport.request("Storage/clear", {});
    }
    list() {
        return this.transport.request("Storage/list", {});
    }
}
class FileSearchService {
    transport;
    constructor(transport) {
        this.transport = transport;
    }
    search(q, opts) {
        return this.transport.request("FileSearch/search", { q, opts });
    }
}
class CommandService {
    transport;
    constructor(transport) {
        this.transport = transport;
    }
    updateCommandMetadata(payload) {
        return this.transport.request("Command/updateCommandMetadata", { payload });
    }
    openExtensionPreferences() {
        return this.transport.request("Command/openExtensionPreferences", {});
    }
    openCommandPreferences() {
        return this.transport.request("Command/openCommandPreferences", {});
    }
}
class OAuthService {
    transport;
    constructor(transport) {
        this.transport = transport;
    }
    authorize(payload) {
        return this.transport.request("OAuth/authorize", { payload });
    }
    getTokens(id) {
        return this.transport.request("OAuth/getTokens", { id });
    }
    setTokens(payload) {
        return this.transport.request("OAuth/setTokens", { payload });
    }
    removeTokens(id) {
        return this.transport.request("OAuth/removeTokens", { id });
    }
}
class EventCoreService {
    transport;
    constructor(transport) {
        this.transport = transport;
    }
    handlerActivated(handler) {
        return this.transport.subscribe("EventCore/handlerActivated", (msg) => handler(msg.id, msg.args));
    }
}
class Client {
    transport;
    constructor(transport) {
        this.transport = transport;
        this.Application = new ApplicationService(this.transport);
        this.UI = new UIService(this.transport);
        this.WindowManagement = new WindowManagementService(this.transport);
        this.Clipboard = new ClipboardService(this.transport);
        this.Storage = new StorageService(this.transport);
        this.FileSearch = new FileSearchService(this.transport);
        this.Command = new CommandService(this.transport);
        this.OAuth = new OAuthService(this.transport);
        this.EventCore = new EventCoreService(this.transport);
    }
    route(msg) {
        this.transport.dispatchMessage(msg);
    }
    Application;
    UI;
    WindowManagement;
    Clipboard;
    Storage;
    FileSearch;
    Command;
    OAuth;
    EventCore;
}
exports.Client = Client;
