// @vicinae/api/dist/api/components/actions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Action = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const node_crypto_1 = require("node:crypto");
const index_1 = require("../hooks/index");
const clipboard_1 = require("../clipboard");
const image_1 = require("../image");
const utils_1 = require("../utils");
const icon_1 = require("../icon");
const controls_1 = require("../controls");
const action_pannel_1 = require("./action-pannel");
const toast_1 = require("../toast");
const ActionRoot = ({ icon, ...props }) => {
    const serializedIcon = icon ? (0, image_1.serializeProtoImage)(icon) : icon;
    const stableIdRef = (0, react_1.useRef)(undefined);
    if (!stableIdRef.current) {
        stableIdRef.current = (0, node_crypto_1.randomUUID)();
    }
    return ((0, jsx_runtime_1.jsx)("action", { ...props, icon: serializedIcon, stableId: stableIdRef.current }));
};
const CopyToClipboard = ({ title = "Copy to clipboard", icon = icon_1.Icon.CopyClipboard, content, concealed = false, onCopy, ...props }) => {
    return ((0, jsx_runtime_1.jsx)(ActionRoot, { title: title, ...props, icon: icon_1.Icon.CopyClipboard, onAction: async () => {
            clipboard_1.Clipboard.copy(content, { concealed });
            (0, controls_1.closeMainWindow)();
            onCopy?.(content);
        } }));
};
const Paste = ({ title = "Paste to active window", icon = icon_1.Icon.CopyClipboard, content, onPaste, ...props }) => {
    return ((0, jsx_runtime_1.jsx)(ActionRoot, { title: title, ...props, icon: icon_1.Icon.CopyClipboard, onAction: async () => {
            (0, controls_1.closeMainWindow)(); // we close before pasting to make sure focus has been properly restored
            clipboard_1.Clipboard.paste(content);
            onPaste?.(content);
        } }));
};
const Open = ({ target, app, ...props }) => {
    return ((0, jsx_runtime_1.jsx)(ActionRoot, { ...props, onAction: () => {
            (0, utils_1.open)(target, app);
        } }));
};
const OpenWith = ({ path, title = "Open with...", shortcut, icon, ...props }) => {
    const [apps, setApps] = (0, react_1.useState)([]);
    const fetchApps = () => {
        (0, utils_1.getApplications)(path).then(setApps);
    };
    return ((0, jsx_runtime_1.jsx)(action_pannel_1.ActionPanel.Submenu, { title: title, icon: icon, shortcut: shortcut, onOpen: () => {
            fetchApps();
        }, children: apps.map((app) => ((0, jsx_runtime_1.jsx)(exports.Action, { title: `Open in ${app.name}`, icon: app.icon, onAction: () => {
                (0, controls_1.closeMainWindow)();
                (0, utils_1.open)(path, app)
                    .then(() => {
                    props.onOpen?.(path);
                })
                    .catch(async (error) => {
                    (0, toast_1.showToast)({ title: `Failed to open app` });
                    console.error("Failed to open app", error);
                });
            } }, app.id))) }));
};
const Trash = ({ title, paths, ...props }) => {
    const actionTitle = title ?? `Delete item${Array.isArray(paths) ? "s" : ""}`;
    return ((0, jsx_runtime_1.jsx)(ActionRoot, { title: actionTitle, ...props, onAction: async () => {
            await (0, utils_1.trash)(paths);
            props.onTrash?.(paths);
        } }));
};
const OpenInBrowser = ({ url, title = "Open in Browser", ...props }) => {
    return ((0, jsx_runtime_1.jsx)(ActionRoot, { ...props, title: title, onAction: async () => {
            await (0, utils_1.open)(url);
            await (0, controls_1.closeMainWindow)();
        } }));
};
const ShowInFinder = ({ path, select = true, title = "Show in file browser", ...props }) => {
    return ((0, jsx_runtime_1.jsx)(ActionRoot, { ...props, title: title, onAction: () => {
            (0, utils_1.showInFileBrowser)(path, { select });
            (0, controls_1.closeMainWindow)();
        } }));
};
const Push = ({ target, ...props }) => {
    const { push } = (0, index_1.useNavigation)();
    return ((0, jsx_runtime_1.jsx)(ActionRoot, { ...props, onAction: () => {
            push(target);
        } }));
};
const SubmitForm = ({ title = "Submit", ...props }) => {
    const stableIdRef = (0, react_1.useRef)(undefined);
    if (!stableIdRef.current) {
        stableIdRef.current = (0, node_crypto_1.randomUUID)();
    }
    const nativeProps = {
        ...props,
        stableId: stableIdRef.current,
        title,
        icon: props.icon ? (0, image_1.serializeProtoImage)(props.icon) : props.icon,
        onAction: () => { },
    };
    return (0, jsx_runtime_1.jsx)("action", { ...nativeProps });
};
// TODO: implement date picker action. This probably requires a full rework of the action panel.
const PickDate = () => {
    return null;
};
const CreateQuicklink = ({ title = "Create Quicklink", quicklink, ...props }) => {
    const stableIdRef = (0, react_1.useRef)(undefined);
    if (!stableIdRef.current) {
        stableIdRef.current = (0, node_crypto_1.randomUUID)();
    }
    const nativeProps = {
        ...props,
        stableId: stableIdRef.current,
        title,
        type: "create-quicklink",
        quicklink: {
            link: quicklink.link,
            name: quicklink.name,
            application: typeof quicklink.application === "string"
                ? quicklink.application
                : quicklink.application?.name,
            icon: quicklink.icon,
        },
        icon: props.icon ? (0, image_1.serializeProtoImage)(props.icon) : props.icon,
        onAction: () => { },
    };
    return (0, jsx_runtime_1.jsx)("action", { ...nativeProps });
};
const RunInTerminal = ({ args, options, icon = icon_1.Icon.Terminal, ...props }) => {
    return ((0, jsx_runtime_1.jsx)(ActionRoot, { ...props, icon: icon, onAction: () => {
            (0, utils_1.runInTerminal)(args, options).then(() => (0, controls_1.closeMainWindow)());
        } }));
};
// TODO: provide actual implementation, one day?
const ToggleQuickLook = (props) => {
    return null;
};
// TODO: provide actual implementation, one day?
const CreateSnippet = (props) => {
    return null;
};
/**
 * @category Actions
 */
exports.Action = Object.assign(ActionRoot, {
    CopyToClipboard,
    Push,
    Open,
    Paste,
    SubmitForm,
    OpenInBrowser,
    OpenWith,
    Trash,
    ShowInFinder,
    RunInTerminal,
    CreateQuicklink,
    ToggleQuickLook,
    CreateSnippet,
    PickDate: Object.assign(PickDate, {
        // TODO: to implement too
        isFullDay: () => false,
    }),
    Style: {
        Regular: "regular",
        Destructive: "destructive",
    },
});
