// @vicinae/api/dist/api/alert.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.confirmAlert = exports.Alert = void 0;
const client_1 = require("./client");
/**
 * @category Alert
 */
var Alert;
(function (Alert) {
    let ActionStyle;
    (function (ActionStyle) {
        ActionStyle["Default"] = "default";
        ActionStyle["Destructive"] = "destructive";
        ActionStyle["Cancel"] = "cancel";
    })(ActionStyle = Alert.ActionStyle || (Alert.ActionStyle = {}));
})(Alert || (exports.Alert = Alert = {}));
const styleMap = {
    [Alert.ActionStyle.Default]: "Default",
    [Alert.ActionStyle.Destructive]: "Destructive",
    [Alert.ActionStyle.Cancel]: "Cancel",
};
/**
 * Display a confirmation dialog with a cancel and confirm options.
 * The action is canceled if the cancel button is pressed or if the alert is dismissed for any reason.
 *
 * @remarks Calling this function when another alert is currently pending will result in the pending alert
 * to be automatically canceled, replacing it with the new one.
 *
 * @returns - whether the action was confirmed.
 *
 * @example
 * ```typescript
 * if (await confirmAlert({ title: 'Are you sure?', description: 'This can't be undone' })) {
 * 	performDangerousAction();
 * }
 * ```
 *
 * @category Alert
 */
const confirmAlert = async (options) => {
    return new Promise((resolve) => {
        (0, client_1.getClient)()
            .UI.confirmAlert({
            title: options.title,
            description: options.message ?? "Are you sure?",
            //icon: options.icon && serializeProtoImage(options.icon),
            rememberUserChoice: false,
            primaryAction: {
                title: options.primaryAction?.title ?? "Confirm",
                style: styleMap[options.primaryAction?.style ?? Alert.ActionStyle.Default],
            },
            dismissAction: {
                title: options.dismissAction?.title ?? "Cancel",
                style: styleMap[options.dismissAction?.style ?? Alert.ActionStyle.Cancel],
            },
        })
            .then((confirmed) => {
            if (confirmed) {
                options.primaryAction?.onAction?.();
            }
            else {
                options.dismissAction?.onAction?.();
            }
            resolve(confirmed);
        });
    });
};
exports.confirmAlert = confirmAlert;
