// @vicinae/api/dist/api/clipboard.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Clipboard = void 0;
const client_1 = require("./client");
/**
 * Access system clipboard and clipboard history features.
 *
 * @category Clipboard
 */
var Clipboard;
(function (Clipboard) {
    function mapContent(content) {
        let ct = {};
        if (typeof content !== "object") {
            ct.text = `${content}`;
        }
        else {
            if (content["file"]) {
                ct.path = content["file"];
            }
            else if (content["html"]) {
                ct.html = content["html"];
                ct.text = content["text"];
            }
            else {
                ct.text = content["text"];
            }
        }
        return ct;
    }
    /**
     * Copy the provided content in the clipboard.
     * The `concealed` option can be passed so that the created clipboard selection
     * does not get indexed by the Vicinae clipboard manager.
     */
    async function copy(text, options = {}) {
        await (0, client_1.getClient)().Clipboard.copy(mapContent(text), {
            concealed: options.concealed ?? false,
        });
    }
    Clipboard.copy = copy;
    /**
     * Paste the provided clipboard content to the active window.
     * If the environment does not support either getting the active window
     * or pasting content to it directly, this will fallback to a regular
     * clipboard copy.
     */
    async function paste(text) {
        await (0, client_1.getClient)().Clipboard.paste(mapContent(text));
    }
    Clipboard.paste = paste;
    /**
     * Read the current content of the clipboard, which can contain text, html and a file path.
     * Note: the offset option is not yet implemented
     *
     * ```ts
     * const { text, html, file } = await Clipboard.read();
     * ```
     */
    async function read() {
        return (0, client_1.getClient)().Clipboard.readContent();
    }
    Clipboard.read = read;
    /**
     * Read the text representation of the current clipboard data. If the data is not text at all, this
     * returns an empty string.
     * If you want to read optional html or file path, consider @see {Clipboard.read}
     * Note: the offset option is not yet implemented.
     *
     * ```ts
     * const text = await Clipboard.readText();
     * ```
     */
    async function readText() {
        return read().then((r) => r.text ?? "");
    }
    Clipboard.readText = readText;
    /**
     * Clear the current clipboard content.
     */
    async function clear() {
        await (0, client_1.getClient)().Clipboard.clear();
    }
    Clipboard.clear = clear;
})(Clipboard || (exports.Clipboard = Clipboard = {}));
