// @vicinae/api/dist/api/controls.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.popToRoot = exports.getSelectedText = exports.clearSearchBar = exports.closeMainWindow = exports.showHUD = exports.PopToRootType = void 0;
const client_1 = require("./client");
/**
 * @category Launcher Window
 */
var PopToRootType;
(function (PopToRootType) {
    /**
     * Translates to Immediate or Suspended depending on the
     * user configuration.
     */
    PopToRootType["Default"] = "default";
    /**
     * Immediately pop to root
     */
    PopToRootType["Immediate"] = "immediate";
    /**
     * Do not pop to root, preserve the navigation state as it is
     */
    PopToRootType["Suspended"] = "suspended";
})(PopToRootType || (exports.PopToRootType = PopToRootType = {}));
const popToRootProtoMap = {
    [PopToRootType.Default]: "Default",
    [PopToRootType.Immediate]: "Immediate",
    [PopToRootType.Suspended]: "Suspended",
};
/**
 * Close the window and show a small HUD where the window was previously opened.
 * Note that the HUD may not be shown in some environments.
 * Similarly to `closeWindow`, it is possible to override the pop to root behavior directly
 * by passing options to this function.
 *
 * @see closeWindow
 *
 * @category Launcher Window
 */
const showHUD = async (title, options) => {
    (0, client_1.getClient)().UI.showHud(title, options?.clearRootSearch ?? false, popToRootProtoMap[options?.popToRootType ?? PopToRootType.Default]);
};
exports.showHUD = showHUD;
/**
 * Close the vicinae launcher window immediately.
 * It is possible to override the `popToRoot` behavior defined in the settings using the options object.
 *
 * @category Launcher Window
 */
const closeMainWindow = async (options = {}) => {
    const { clearRootSearch = false, popToRootType = PopToRootType.Default } = options;
    await (0, client_1.getClient)().UI.closeMainWindow(clearRootSearch, popToRootProtoMap[popToRootType]);
};
exports.closeMainWindow = closeMainWindow;
/**
 * @category Launcher Window
 */
const clearSearchBar = async () => {
    await (0, client_1.getClient)().UI.setSearchText("");
};
exports.clearSearchBar = clearSearchBar;
/**
 * Get the text that is currently selected by the user.
 * How this is implemented depends on the environment but all it does is usually
 * read the clipboard's primary selection buffer.
 *
 * @category Launcher Window
 */
const getSelectedText = async () => {
    return (0, client_1.getClient)().UI.getSelectedText();
};
exports.getSelectedText = getSelectedText;
/**
 * Pop to the root of the navigation stack, optionally clearing the search bar.
 *
 * @category Launcher Window
 */
const popToRoot = async (options) => {
    await (0, client_1.getClient)().UI.popToRoot(options?.clearSearchBar ?? false);
};
exports.popToRoot = popToRoot;
