// @vicinae/api/dist/api/image.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.serializeProtoImage = exports.Image = void 0;
const color_1 = require("./color");
const icon_1 = require("./icon");
/**
 * @category Image
 */
var Image;
(function (Image) {
    let Mask;
    (function (Mask) {
        Mask["Circle"] = "circle";
        Mask["RoundedRectangle"] = "roundedRectangle";
    })(Mask = Image.Mask || (Image.Mask = {}));
})(Image || (exports.Image = Image = {}));
const isFileIcon = (v) => typeof v === "object" &&
    v !== null &&
    typeof v.fileIcon === "string";
const maskMap = {
    [Image.Mask.Circle]: "Circle",
    [Image.Mask.RoundedRectangle]: "RoundedRectangle",
};
const serializeProtoImage = (image) => {
    const serializeSource = (payload) => {
        if (typeof payload === "object") {
            const tmp = payload;
            return {
                themed: { light: tmp.light?.toString(), dark: tmp.dark?.toString() },
            };
        }
        return { raw: payload.toString() };
    };
    if (image instanceof URL || typeof image === "string") {
        return { source: { raw: image.toString() } };
    }
    if (isFileIcon(image)) {
        return { fileIcon: image.fileIcon };
    }
    const img = image;
    // img.source should technically not be null, but it somehow still happens at times with some
    // raycast extensions
    if (!img.source)
        return (0, exports.serializeProtoImage)(icon_1.Icon.QuestionMarkCircle);
    return {
        source: serializeSource(img.source),
        fallback: img.fallback ? serializeSource(img.fallback) : undefined,
        mask: img.mask ? maskMap[img.mask] : undefined,
        tintColor: img.tintColor ? (0, color_1.serializeColorLike)(img.tintColor) : undefined,
    };
};
exports.serializeProtoImage = serializeProtoImage;
