// @vicinae/api/dist/api/local-storage.js
"use strict";
// Implementation of Raycast's storage API: https://developers.raycast.com/api-reference/storage
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalStorage = void 0;
const client_1 = require("./client");
/**
 * @category Local Storage
 */
var LocalStorage;
(function (LocalStorage) {
    async function getItem(key) {
        return (0, client_1.getClient)().Storage.get(key);
    }
    LocalStorage.getItem = getItem;
    async function setItem(key, value) {
        await (0, client_1.getClient)().Storage.set(key, value);
    }
    LocalStorage.setItem = setItem;
    async function removeItem(key) {
        await (0, client_1.getClient)().Storage.remove(key);
    }
    LocalStorage.removeItem = removeItem;
    async function allItems() {
        return (0, client_1.getClient)().Storage.list();
    }
    LocalStorage.allItems = allItems;
    async function clear() {
        await (0, client_1.getClient)().Storage.clear();
    }
    LocalStorage.clear = clear;
})(LocalStorage || (exports.LocalStorage = LocalStorage = {}));
