// @vicinae/api/dist/api/toast.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.showToast = exports.Toast = void 0;
const node_crypto_1 = require("node:crypto");
const client_1 = require("./client");
/**
 * A Toast with a certain style, title, and message.
 *
 * @example
 * ```typescript
 * import { showToast, Toast } from "@raycast/api";
 * import { setTimeout } from "timers/promises";
 *
 * export default async () => {
 *   const toast = await showToast({ style: Toast.Style.Animated, title: "Uploading image" });
 *
 *   await setTimeout(1000);
 *
 *   toast.style = Toast.Style.Success;
 *   toast.title = "Uploaded image";
 * };
 * ```
 *
 * @category Toast
 */
class Toast {
    options;
    callbacks = {};
    id;
    styleMap = {
        [Toast.Style.Success]: "Success",
        [Toast.Style.Failure]: "Error",
        [Toast.Style.Animated]: "Dynamic",
    };
    /**
     * Deprecated - Use `showToast` instead
     */
    constructor(props) {
        this.id = `toast_${(0, node_crypto_1.randomBytes)(16).toString("hex")}`;
        this.options = {
            title: props.title,
            style: props.style ?? Toast.Style.Success,
            message: props.message,
        };
        /*
        if (props.primaryAction) {
            const { onAction } = props.primaryAction;
            const { id } = bus.addEventHandler(() => onAction(this));

            //this.options.primaryAction = props.primaryAction;
            this.callbacks.primary = id;
        }

        if (props.secondaryAction) {
            const { onAction } = props.secondaryAction;
            const { id } = bus.addEventHandler(() => onAction(this));

            //this.options.secondaryAction = props.secondaryAction;
            this.callbacks.secondary = id;
        }
        */
    }
    /**
     * The style of a Toast.
     */
    get style() {
        return this.options.style;
    }
    set style(style) {
        this.options.style = style;
        this.update();
    }
    /**
     * The title of a Toast. Displayed on the top.
     */
    get title() {
        return this.options.title;
    }
    set title(title) {
        this.options.title = title;
        this.update();
    }
    /**
     * An additional message for the Toast. Useful to show more information, e.g. an identifier of a newly created asset.
     */
    get message() {
        return this.options.message;
    }
    set message(message) {
        this.options.message = message;
    }
    /**
     * The primary Action the user can take when hovering on the Toast.
     */
    /*
    get primaryAction(): Toast.ActionOptions | undefined {
        return this.options.primaryAction;
    }
    set primaryAction(action: Toast.ActionOptions | undefined) {
        this.options.primaryAction = action;
    }
    get secondaryAction(): Toast.ActionOptions | undefined {
        return this.options.secondaryAction;
    }
    set secondaryAction(action: Toast.ActionOptions | undefined) {
        this.options.secondaryAction = action;
    }
    */
    async update() {
        await (0, client_1.getClient)().UI.showToast(this.id, this.title, this.message ?? "", this.styleMap[this.style]);
    }
    /**
     * Shows the Toast.
     *
     * @returns A Promise that resolves when the toast is shown.
     */
    async show() {
        this.update();
    }
    /**
     * Hides the Toast.
     *
     * @returns A Promise that resolves when toast is hidden.
     */
    async hide() {
        await (0, client_1.getClient)().UI.hideToast(this.id);
    }
}
exports.Toast = Toast;
/**
 * @category Toast
 */
(function (Toast) {
    /**
     * Defines the visual style of the Toast.
     *
     * @remarks
     * Use {@link Toast.Style.Success} for confirmations and {@link Toast.Style.Failure} for displaying errors.
     * Use {@link Toast.Style.Animated} when your Toast should be shown until a process is completed.
     * You can hide it later by using {@link Toast.hide} or update the properties of an existing Toast.
     */
    let Style;
    (function (Style) {
        Style["Success"] = "success";
        Style["Failure"] = "failure";
        Style["Animated"] = "animated";
    })(Style = Toast.Style || (Toast.Style = {}));
})(Toast || (exports.Toast = Toast = {}));
/**
 * Show a toast notification on the bottom left of the Vicinae window, briefly replacing the navigation title.
 * Toast can have different style depending on the nature of the notification, and can even use `Toast.Style.Animated`
 * to display a loading spinner and refresh the title in realtime.
 *
 * @example
 * ```typescript
 * import { showToast, Toast } from "@raycast/api";
 * import { setTimeout } from "timers/promises";
 *
 * export default async () => {
 *   const toast = await showToast({ style: Toast.Style.Animated, title: "Uploading image" });
 *
 *   await setTimeout(1000);
 *
 *   toast.style = Toast.Style.Success;
 *   toast.title = "Uploaded image";
 * };
 * ```
 *
 * @category Toast
 */
const showToast = async (init, title = "", message) => {
    const toast = typeof init === "string"
        ? new Toast({ style: init, message, title })
        : new Toast(init);
    await toast.show();
    return toast;
};
exports.showToast = showToast;
