// @vicinae/api/dist/api/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sendDesktopNotification = exports.showInFileBrowser = exports.getDefaultApplication = exports.getApplications = exports.getFrontmostApplication = exports.open = exports.runInTerminal = exports.trash = void 0;
const promises_1 = require("node:fs/promises");
const window_management_1 = require("./window-management");
const client_1 = require("./client");
const image_1 = require("./image");
/**
  @ignore - we should probably move this to raycast compat, I don't think we want that.
 */
const trash = async (path) => {
    const targets = Array.isArray(path) ? path : [path];
    const promises = targets.map((p) => (0, promises_1.rm)(p, { recursive: true }));
    await Promise.all(promises);
};
exports.trash = trash;
/**
 * Run a command in a new terminal emulator window.
 *
 * @param args - the command line to execute. This is *not* getting interpreted by a shell.
 * @param options - list of options that can be passed in order to tweak the behavior of the terminal window.
 *
 * @example
 * ```typescript
 * await runInTerminal(['journalctl', '--user', '-u', '-f', 'vicinae']);
 * // or, inside a shell:
 * await runInTerminal(['/bin/bash', '-c', 'echo "dis is my home: $HOME"'], { hold: true });
 * ```
 *
 * @category System
 */
const runInTerminal = async (args, options = {}) => {
    const { hold = false, appId, title, workingDirectory } = options;
    await (0, client_1.getClient)().Application.runInTerminal({
        cmdline: args,
        hold,
        appId,
        title,
        workingDirectory,
    });
};
exports.runInTerminal = runInTerminal;
/**
 * @category System
 */
const open = async (target, app) => {
    let appId;
    if (app) {
        if (typeof app === "string") {
            appId = app;
        }
        else {
            appId = app.id;
        }
    }
    await (0, client_1.getClient)().Application.open(target, appId);
};
exports.open = open;
/**
 * @category System
 */
const getFrontmostApplication = async () => {
    const { application } = await window_management_1.WindowManagement.getActiveWindow();
    if (!application) {
        throw new Error(`Could not get frontmost application`);
    }
    return application;
};
exports.getFrontmostApplication = getFrontmostApplication;
/**
 * @category System
 */
const getApplications = async (target) => {
    return (0, client_1.getClient)().Application.list(target);
};
exports.getApplications = getApplications;
/**
 * @category System
 */
const getDefaultApplication = async (path) => {
    return (0, client_1.getClient)().Application.getDefault(path);
};
exports.getDefaultApplication = getDefaultApplication;
/**
 * @category System
 */
const showInFileBrowser = async (path, options = {}) => {
    await (0, client_1.getClient)().Application.showInFileBrowser(path.toString(), options.select ?? true);
};
exports.showInFileBrowser = showInFileBrowser;
/**
 * @category System
 */
const sendDesktopNotification = (payload) => (0, client_1.getClient)().UI.sendDesktopNotification({
    title: payload.title,
    body: payload.body,
    icon: payload.icon ? (0, image_1.serializeProtoImage)(payload.icon) : undefined,
    urgency: payload.urgency ?? "Normal",
});
exports.sendDesktopNotification = sendDesktopNotification;
