// @vicinae/api/dist/api/window-management.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WindowManagement = void 0;
const client_1 = require("./client");
const transformWorkspace = (proto) => {
    return {
        id: proto.id,
        name: proto.name,
        active: proto.active,
        monitorId: proto.monitor,
    };
};
const transformWindow = (proto) => {
    return {
        id: proto.id,
        title: proto.title,
        workspaceId: proto.workspaceId,
        active: proto.active,
        bounds: {
            position: { x: proto.bounds.x, y: proto.bounds.y },
            size: { width: proto.bounds.width, height: proto.bounds.height },
        },
        application: proto.app,
        focus() {
            return WindowManagement.focusWindow(this);
        },
    };
};
/**
 * Access Vicinae's window management features.
 *
 * @remarks
 * Window management support varies a lot depending on the environment.
 * Right now it is pretty well supported on almost all linux desktop environments except KDE.
 *
 * @example
 * ```typescript
 * import { WindowManagement } from '@vicinae/api';
 *
 * const wins = await WindowManagement.getWindows();
 * const browserWindow = wins.find(w => w.application?.name?.includes('firefox'));
 *
 * if (browserWindow) {
 *  await browserWindow.focus();
 * }
 * ```
 *
 * @category Window Management
 * @public
 */
var WindowManagement;
(function (WindowManagement) {
    async function getWindows(workspaceId) {
        return (0, client_1.getClient)()
            .WindowManagement.getWindows(workspaceId)
            .then((wins) => wins.map(transformWindow));
    }
    WindowManagement.getWindows = getWindows;
    /**
     * Focus `window`.
     *
     * @remarks
     * Window objects have a {@link Window.focus} method that can be used to achieve the same thing on a specific window directly.
     *
     * @param window - the window to focus. You may want to make sure this window still exists when you request focus.
     *
     * @return `true` if the window was focused, `false` otherwise.
     * A window may not have been focused because it doesn't accept focus (e.g some layer shell surfaces)
     * or simply because it doesn't exist anymore.
     *
     * @see {@link Window.focus}
     */
    async function focusWindow(window) {
        return (0, client_1.getClient)().WindowManagement.focusWindow(window.id);
    }
    WindowManagement.focusWindow = focusWindow;
    /**
     * Return the list of screens (physical and virtual) currently attached to the computer.
     */
    async function getScreens() {
        return (0, client_1.getClient)()
            .WindowManagement.getScreens()
            .then((screens) => screens.map((sc) => ({
            name: sc.name,
            make: sc.make,
            model: sc.model,
            serial: sc.serial,
            bounds: {
                position: { x: sc.bounds.x, y: sc.bounds.y },
                size: { width: sc.bounds.width, height: sc.bounds.height },
            },
        })));
    }
    WindowManagement.getScreens = getScreens;
    async function getActiveWorkspace() {
        return (0, client_1.getClient)()
            .WindowManagement.getActiveWorkspace()
            .then(transformWorkspace);
    }
    WindowManagement.getActiveWorkspace = getActiveWorkspace;
    async function getWorkspaces() {
        return (0, client_1.getClient)()
            .WindowManagement.getWorkspaces()
            .then((workspaces) => workspaces.map(transformWorkspace));
    }
    WindowManagement.getWorkspaces = getWorkspaces;
    async function getWindowsOnActiveWorkspace() {
        return (0, client_1.getClient)()
            .WindowManagement.getActiveWorkspace()
            .then((ws) => getWindows(ws.id));
    }
    WindowManagement.getWindowsOnActiveWorkspace = getWindowsOnActiveWorkspace;
    async function getActiveWindow() {
        return (0, client_1.getClient)().WindowManagement.getActiveWindow().then(transformWindow);
    }
    WindowManagement.getActiveWindow = getActiveWindow;
})(WindowManagement || (exports.WindowManagement = WindowManagement = {}));
